### PR TITLE
[lora, perf] refactor: avoid one-hot routing in eager MoE LoRA

### PR DESCRIPTION
**Pull request by Levius-Fubuki**
### What does this PR do?

The eager shared and independent MoE-LoRA paths currently materialize a
`[num_tokens, top_k, num_experts]` one-hot tensor, reduce it to find active
experts, and then call `torch.where` for every active expert.

This PR groups the `[num_tokens, top_k]` routing assignments directly:

- transpose to preserve the original top-k-major traversal order;
- stable-sort the flattened assignments by expert ID;
- split consecutive assignments by expert;
- recover `(top_k_pos, token_idx)` from each flattened position.

This reduces the routing workspace from `O(T * K * E)` to `O(T * K)`, while
preserving the original per-expert assignment order and leaving the projection
and routing-weight math unchanged.

### Checklist Before Starting

- Search for relative PRs/issues and link here: no overlapping PR or issue was
  found. #758 introduced MoE-LoRA, but does not contain this optimization.
- This PR title follows the required format:
  `[lora, perf] refactor: avoid one-hot routing in eager MoE LoRA`.
- This change is scoped to the existing eager MoE-LoRA implementation.

#### Regression tests

```bash
PYTHONPATH=. python -m pytest -q tests/lora/test_moe_lora_eager.py
```

Result: `37 passed`.

The new tests:

- compare the grouped assignments against the previous one-hot implementation
  for top-k 1, top-k 2, repeated/unused experts, and a random top-k 8 case;
- exercise both shared and independent Qwen3-MoE LoRA wrappers;
- fail if either eager forward path calls `torch.nn.functional.one_hot`.

#### Lint and formatting

```bash
ruff format --check veomni/lora/moe_layers.py tests/lora/test_moe_lora_eager.py
ruff check veomni/lora/moe_layers.py tests/lora/test_moe_lora_eager.py
git diff --check HEAD^
```

All checks pass.

#### RTX 4090 microbenchmark

Environment: RTX 4090, CUDA 13.0, PyTorch 2.11.0+cu130. Measurements use CUDA
events after warmup. The routing-only benchmark covers 48 combinations of
tokens (`128` to `8192`), top-k (`1`, `2`, `4`, `8`), and experts (`8` to
`256`).

| Benchmark | Previous | This PR | Change |
| --- | ---: | ---: | ---: |
| Routing-only median across 48 cases | — | — | 48/48 faster; median `-58.3%` |
| Routing-only, T=8192, K=8, E=256 | 15.5867 ms | 5.7285 ms | `-63.2%` |
| Peak routing memory, T=8192, K=8, E=256 | 329.565 MiB | 13.619 MiB | `-95.9%` |
| Shared wrapper forward | 6.923 ms | 5.395 ms | `-22.1%` |
| Independent wrapper forward | 11.153 ms | 7.434 ms | `-33.3%` |
| Independent wrapper forward + backward | 45.479 ms | 41.324 ms | `-9.1%` |

The shared forward + backward result was clock-sensitive across separate
processes: one run measured `+0.9%` and a repeat measured `-8.8%`. The
same-process routing-only result is therefore the primary performance signal.

### API and Usage Example

No public API or configuration changes.

### Design & Code Changes

- Add `_group_routing_assignments` as the shared eager routing helper.
- Replace one-hot mask construction in `LoraSharedExperts` and
  `LoraIndependentExperts` with grouped assignment iteration.
- Add equivalence and no-one-hot regression coverage.

### Checklist Before Submitting

- [x] Read the contribution guide.
- [x] Added regression coverage for both eager MoE-LoRA modes.
- [x] Ran focused formatting, lint, and test checks.
- [x] No documentation update is required because the public API is unchanged.
- [x] The tests live in the existing `tests/lora/test_moe_lora_eager.py` suite.

### Limitations

- The benchmark was run on a single RTX 4090.
- This only changes the eager MoE-LoRA paths; fused MoE kernels are unaffected.

### AI assistance

AI assistance was used to explore the optimization, implement the patch, and
draft tests and documentation. I reviewed the changes and validated them with
the tests and benchmarks above.

## Summary by CodeRabbit

- **Bug Fixes**
  - Fixed eager Mixture-of-Experts LoRA routing to ignore padded expert entries and process only valid experts.
  - Preserved correct routing behavior across shared- and independent-expert modes.

- **Performance**
  - Reduced temporary routing overhead by avoiding materialization of one-hot routing tensors during eager execution.

- **Tests**
  - Added coverage for routing assignments, padded expert handling, and eager forward execution across supported LoRA modes.

### CODE REVIEW

**Review by Coach257 — APPROVED** [review]

File: veomni/lora/moe_layers.py [hunk 1]
```diff
@@ -164,6 +164,20 @@
...
+    """Yield expert routing groups in the eager path's top-k-major order."""
+    num_tokens = top_k_index.shape[0]
+    flat_experts = top_k_index.T.reshape(-1)
+    sorted_experts, flat_positions = torch.sort(flat_experts, stable=True)
+    expert_ids, counts = torch.unique_consecutive(sorted_experts, return_counts=True)
+
+    offset = 0
+    for expert_idx, count in zip(expert_ids.tolist(), counts.tolist(), strict=True):
+        group_positions = flat_positions[offset : offset + count]
+        yield expert_idx, group_positions // num_tokens, group_positions % num_tokens
```
[the remark this answers is not in the document]
Comment (Coach257) [reply veomni/lora/moe_layers.py]: @Levius-Fubuki Please fix this

[the remark this answers is not in the document]
Comment (Levius-Fubuki) [reply veomni/lora/moe_layers.py]: Fixed in d7c8367. `_group_routing_assignments` now receives `num_experts` and skips the padding sentinel before either eager path indexes expert tensors. Added direct helper coverage and shared/independent eager forward regression coverage. Ruff formatting, lint, and `git diff --check` pass; the GPU test suite is rerunning in CI.
